google/cloud/bigquery/job/query.py

```diff
@@ -988,15 +988,25 @@ def done(self, retry=DEFAULT_RETRY, timeout=None, reload=True):
                 unfinished jobs before checking. Default ``True``.
 
         Returns:
-            bool: True if the job is complete, False otherwise.
+            bool: True if the job is complete, False otherwise. If job status
+                cannot be determined due to a non-retryable error or too many
+                retries, the method returns ``True``.
         """
         # Do not refresh if the state is already done, as the job will not
         # change once complete.
         is_done = self.state == _DONE_STATE
         if not reload or is_done:
             return is_done
 
-        self._reload_query_results(retry=retry, timeout=timeout)
+        try:
+            self._reload_query_results(retry=retry, timeout=timeout)
+        except Exception as exc:
+            # The job state might still be "RUNNING", but if an exception bubbles
+            # up (either a RetryError or a non-retryable error), we declare that we
+            # are done for good (and the blocking pull should not continue polling).
+            # Ditto in a similar try-except block below.
+            self.set_exception(exc)
+            return True
```

Comment:
Reviewers: Please confirm that returning `True` is fine in such case.

If returning `self.state == _DONE_STATE`, the query from the issue description can cause a long retry loop, as the backend seems to always return the state `RUNNING` for the job, even though an error in the query can (presumably?) be detected immediately and there's no need to keep the query running.

Reply:
I think we need to reload the job object in this case. `jobs.get` has different error conditions than `jobs.getQueryResults`.

Reply:
That is, `jobs.get` (i.e. `self.reload()`) should return successfully, even for failed jobs.

Reply:
IIRC I tried that, but the state was still "RUNNING". I can check once more tomorrow, just in case.

Should `self.reload()` indeed not work, what would be the preferred backup plan?